### PR TITLE
Add podman remote support

### LINKED ISSUE #27: Add support for `podman --remote`
## Please describe what you would like to see in this project

Podman has the ability to run containers on a remote system via a socket connected to over ssh. We would like the ability to utilize this feature with both the default deployer and within a plugin step.

## Please describe your use case

Running workflow plugins on target remote systems provides flexibility to include non-k8s remote targets, which would be helpful for CI execution as well as for parallelization of workflows across multiple targets.

## Additional context

The server-side prerequisite for this is for the remote target host to enable the socket:
```
systemctl [--user] --now enable podman.socket
```
You can validate this is working on the server side with:
```
podman --remote info
```

On the client side (from where the workflow is run), an ssh identity for connection authentication with the remote host is required.

From the client, the remote connection is setup with:
```
podman system connection add <connection_name> --identity <path_to_ssh_private_key> ssh://<user>@<remote_host>/run/user/<uid>/podman/podman.sock
```

This connection can be validated on the client side with:
```
podman system connection list
podman --connection <connection_name> info
```

Once the remote connection is established and validated, a `podman` command can be run as usual, adding only `--connection <connection_name>` to the parameters (before the `run` command), for example:
```
podman --connection <connection_name> run -i --rm quay.io/arcalot/arcaflow-plugin-example:latest --schema
```

### PR DESCRIPTION
**Pull request by ewchong**
Close #27

## Changes introduced with this PR

* Add support for running remote podman commands using [`--connection`](https://docs.podman.io/en/latest/markdown/podman.1.html#connection-c)

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: connector.go [hunk 1]
```diff
@@ -36,7 +36,6 @@ func (c *Connector) Deploy(ctx context.Context, image string) (deployer.Plugin,
 		c.logger.Errorf("oops, neither podman -> path provided in configuration nor binary found in $PATH")
 		panic("oops, neither podman -> path provided in configuration nor binary found in $PATH")
 	}
-
```
Comment (jaredoconnell) [inline connector.go:39]: This should probably be reverted unless there was a reason for it.

File: internal/cliwrapper/cliwrapper.go [hunk 2]
```diff
@@ -15,12 +15,19 @@ import (
 type cliWrapper struct {
 	podmanFullPath string
 	logger         log.Logger
+	runtimeContext []string
 }
 
-func NewCliWrapper(fullPath string, logger log.Logger) CliWrapper {
+func NewCliWrapper(fullPath string, logger log.Logger, context string) CliWrapper {
+	// Prepand the podman argument
```
Comment (dbutenhof) [inline internal/cliwrapper/cliwrapper.go:22]: "Prepend" ... although that's not an accurate description of either the next line(s) or of the function. This constructs a new CLI wrapper object, and you've added the capability to specify a non-default connection parameter. (Which will, in time, become a podman option.)

File: internal/cliwrapper/cliwrapper.go [hunk 3]
```diff
@@ -15,12 +15,19 @@ import (
 type cliWrapper struct {
 	podmanFullPath string
 	logger         log.Logger
+	runtimeContext []string
```
Comment (dbutenhof) [inline internal/cliwrapper/cliwrapper.go:18]: Why `runtimeContext` ... that seems really vague and generic when what you're specifying is a remote podman "connection" name.

File: internal/cliwrapper/cliwrapper.go [hunk 4]
```diff
@@ -15,12 +15,19 @@ import (
...
 }
 
-func NewCliWrapper(fullPath string, logger log.Logger) CliWrapper {
+func NewCliWrapper(fullPath string, logger log.Logger, context string) CliWrapper {
+	// Prepand the podman argument
+	runtimeContext := []string{}
+	if context != "" {
+		runtimeContext = append([]string{"--connection"}, context)
```
Comment (dbutenhof) [inline internal/cliwrapper/cliwrapper.go:25]: Although `--connection <name>` is a conventional CLI option representation, a single string `--connection=<name>` should work exactly the same (unless `podman` is using an unconventional parser), and handling this additional option as a single string would simplify your logic.

File: internal/cliwrapper/cliwrapper.go (hunk 3 above)
Comment (ewchong) [reply internal/cliwrapper/cliwrapper.go:18]: Making it generic to keep the same name in the docker deployer which has a similar command: https://docs.docker.com/engine/context/working-with-contexts/#use-a-different-context

File: internal/cliwrapper/cliwrapper.go (hunk 2 above)
Comment (ewchong) [reply internal/cliwrapper/cliwrapper.go]: Ah this was a copy/paste relic from a previous failed implementation, updated.

File: internal/cliwrapper/cliwrapper.go (hunk 4 above)
Comment (ewchong) [reply internal/cliwrapper/cliwrapper.go]: Updated

File: internal/cliwrapper/cliwrapper.go (hunk 3 above)
Comment (jaredoconnell) [reply internal/cliwrapper/cliwrapper.go:18]: So the question is then can the same type of input be used for both podman and docker?

Comment (dbutenhof) [reply internal/cliwrapper/cliwrapper.go:18]: Ah: and Podman seems to support `podman context` as well as `podman system connection`. Still, I think "runtime context" isn't obviously the same, and sounds a lot less specific. If you want to hold onto the Docker "context" perhaps try to "connect" that to the Podman "connection" term as something like `ConnectionContext` (which actually gives some clue to what it does).

And as hinted by Jared's comment, we need to think about whether to extend the podman deployer support to the generalized "connection/context" mechanism without doing the same for the docker deployer. It should definitely look the same to the user.

Comment (ewchong) [reply internal/cliwrapper/cliwrapper.go:18]: Did not realize podman had a context option, looks like they are aliases for podman connection based on this https://github.com/containers/podman/issues/12811#issuecomment-1153128175. Could get rid of `connection` and use `context` so it applies to both.

File: internal/cliwrapper/cliwrapper_test.go [hunk 5]
```diff
@@ -52,7 +52,7 @@ func TestPodman_PullImage(t *testing.T) {
 	logger := log.NewTestLogger(t)
 	tests.RemoveImage(logger, tests.TestImageMultiPlatform)
 
-	podman := NewCliWrapper(tests.GetPodmanPath(), logger)
+	podman := NewCliWrapper(tests.GetPodmanPath(), logger, "")
```
Comment (webbnh) [inline internal/cliwrapper/cliwrapper_test.go:55]: I think we should try to come up with a test which exercises this with a non-empty third argument.

**Review by webbnh — CHANGES_REQUESTED** [review]
Thanks Ed for bringing this forward.

Before we merge this, I think the invocation of the Podman command needs to be further encapsulated (and I'll work up a PR for your branch for that).

Also, I think we should try to come up with a way to test this new code with something other than an empty value.  (I'll ponder that, but please do as well.)

**Review by webbnh — CHANGES_REQUESTED** [review]
Also, the CI is unhappy with you:
- [Go is feeling critical](https://github.com/arcalot/arcaflow-engine-deployer-podman/actions/runs/8838926589/job/24305315263?pr=61#step:4:6097); and,
- [`podman` is not happy with you, either.](https://github.com/arcalot/arcaflow-engine-deployer-podman/actions/runs/8838926589/job/24305314846?pr=61#step:6:27)

**Comment by webbnh** [comment]
> the CI is unhappy with you:

@ewchong, don't be in too much hurry to chase these -- I think my refactor will hit all of them.

File: internal/cliwrapper/cliwrapper.go [hunk 6]
```diff
@@ -15,12 +15,18 @@ import (
...
+	// Specify podman --connection string if provided
+	if context != "" {
+		context = "--connection=" + context
+	}
 	return &cliWrapper{
 		podmanFullPath: fullPath,
 		logger:         logger,
+		runtimeContext: []string{context},
 	}
 }
```
Comment (webbnh) [inline internal/cliwrapper/cliwrapper.go:31]: This produces an awkward result when `context` is an empty string:  it adds an empty argument to the Podman command line which produces weird effects (like making the relevant tests fail!  🙃).

I've [coded around that](https://github.com/ewchong/arcaflow-engine-deployer-podman/pull/1) (by not adding the context when it would contribute an empty argument), but it's kind of ugly, and it would be better if there were some other approach (like leaving the `runtimeContext` slice _empty_, instead of putting an empty string into it).

**Review by webbnh — COMMENTED** [review]
I've posted https://github.com/ewchong/arcaflow-engine-deployer-podman/pull/1 which encapsulates the generation of the Podman command (so you only have to add the runtime context in a single place), and it also addresses the lint and test problems.

However, my changes include a workaround for an issue in your changes which I'm hoping that you'll address (this was the source of the test failures; see my comment below).  And, I still think we need to come up with a way to actually test your changes.  🙂

File: internal/cliwrapper/cliwrapper.go (hunk 3 above)
Comment (dbutenhof) [reply internal/cliwrapper/cliwrapper.go:18]: Although I haven't looked into all the details, I believe that the original docker "context" (they don't reference "runtime context", which I still think implies something very different) deals with mapping remote sockets to a "context name", a capability that podman (I believe in 4.3) adopted as an extension to their "connection" mechanism which allowed specifying a REST API socket, including using designated SSH RSA tokens, which seems to go a bit beyond docker; while the podman "context" models the docker CLI syntax and accepts (but apparently doesn't implement) some of the options...

So basically "docker context" and "podman context" allow specifying a remote port, including ssh, tcp, or unix sockets, while podman "connections" add ssh RSA tokens for ID.

File: internal/cliwrapper/cliwrapper.go [hunk 7]
```diff
@@ -15,12 +15,18 @@ import (
...
 	logger         log.Logger
+	runtimeContext []string
 }
 
-func NewCliWrapper(fullPath string, logger log.Logger) CliWrapper {
+func NewCliWrapper(fullPath string, logger log.Logger, context string) CliWrapper {
+	// Specify podman --connection string if provided
+	if context != "" {
+		context = "--connection=" + context
+	}
```
[the remark this answers is not in the document]
Comment (dbutenhof) [reply internal/cliwrapper/cliwrapper.go:25]: Is this actually a connection "destination path" (e.g., a literal `ssh://root@server.fubar.com:22` or a "connection name" associated with a destination path through `podman system connection add` -- or can it be either? I've been reading so much documentation today for various PRs my brain is losing context/connection, and the `podman --help` just says "Connection to use"...

[the remark this answers is not in the document]
Comment (webbnh) [reply internal/cliwrapper/cliwrapper.go:25]: I'm wondering the same thing.  The `podman-remote` documentation shows `--connection=name` which makes me think that, if you want to specify a URL, you have to specify `--url` instead.

[the remark this answers is not in the document]
Comment (webbnh) [reply internal/cliwrapper/cliwrapper.go:25]: When I did this stuff for Pbench, I used the `CONTAINER_HOST` environment variable (which takes the URL) -- I stayed away from the `--connection` option.  😇

File: internal/cliwrapper/cliwrapper.go [hunk 8]
```diff
@@ -93,22 +81,45 @@ func (p *cliWrapper) Deploy(image string, podmanArgs []string, containerArgs []s
...
 		p.logger.Infof("successfully removed container %s", containerName)
 	}
 	return nil
 }
+
+func (p *cliWrapper) getPodmanCmd(cmdArgs ...string) *exec.Cmd {
+	var commandArgs []string
+	if len(p.runtimeContext) > 0 && p.runtimeContext[0] != "" {
+		commandArgs = append(commandArgs, p.runtimeContext...)
+	}
```
Comment (webbnh) [inline internal/cliwrapper/cliwrapper.go:106]: It would be better if this code only had to check one thing instead of two.  For instance, if `NewCliWrapper()` ensured that `p.runtimeContext` contained _no entries_ when there were no options to provide, then we could reduce this code to just:
```go
	commandArgs := p.runtimeContext
```
Because `p.runtimeContext` would either be an empty list or a list with the stuff that we already want -- either way, we wouldn't need the conditional.

**Review by webbnh — APPROVED** [review]
This looks good, now, Ed (if I do say so myself 😉) -- the tests are all passing and your additions are nicely localized.

However, speaking of tests, is there some way that we can set up a context for the local host so that we can test the changes you've made here with an actual `--connection` in the invocation?

File: internal/cliwrapper/cliwrapper.go (hunk 6 above)
Comment (webbnh) [reply internal/cliwrapper/cliwrapper.go:31]: Continued [below](https://github.com/arcalot/arcaflow-engine-deployer-podman/pull/61#discussion_r1583655416).

File: internal/cliwrapper/cliwrapper.go (hunk 3 above)
Comment (ewchong) [reply internal/cliwrapper/cliwrapper.go:18]: I tried using the `--context` option on podman and it seems half baked. It doesn't allow you to add a SSH key so I plan on making this a podman only option and make the variables more specific.

Comment (ewchong) [reply internal/cliwrapper/cliwrapper.go]: Tracking it in this thread instead: https://github.com/arcalot/arcaflow-engine-deployer-podman/pull/61#discussion_r1581211823

File: internal/cliwrapper/cliwrapper.go (hunk 8 above)
Comment (ewchong) [reply internal/cliwrapper/cliwrapper.go]: Made the above suggested changes and added the check here:
https://github.com/arcalot/arcaflow-engine-deployer-podman/pull/61/commits/eb6cd6347bb2e27d49d5797138c2569f9910aa78#diff-1123427b979173bab05da57bb98e3033f246de79c9d274320e84fd4f3b22e68bR24-R25

File: internal/cliwrapper/cliwrapper_test.go (hunk 5 above)
Comment (dbutenhof) [reply internal/cliwrapper/cliwrapper_test.go:55]: I was kinda curious, so I experimented a bit yesterday and couldn't figure out how to set a `podman` system connection to the local remote socket, but I figured it out this morning:

```bash
$ systemctl --user start podman.socket
$ podman system connection add test unix:///run/user/${UID}/podman/podman.sock
$ podman --connection=test ps
CONTAINER ID  IMAGE                      COMMAND         CREATED       STATUS       PORTS                   NAMES
e3ee2b57514a  localhost/backend:latest                   39 hours ago  Up 39 hours                          back
2e2052ba3580  localhost/frontend:latest  yarn run start  39 hours ago  Up 39 hours  0.0.0.0:3001->3000/tcp  front
```

Interestingly, the global `/run/podman/podman.sock` doesn't work because I "can't read it".

Comment (ewchong) [reply internal/cliwrapper/cliwrapper_test.go:55]: I came to the same conclusion last night though since my account was root, was able to access the `/run/podman/podan.sock`. 

Couple caveats I see:
- It sets that as your default connection if you don't have one at all, probably not a big deal. This seems to be the case on my linux test machine. 
- It won't work on macOS but the upside is does have connections already present that can be used:
```
▶  podman system connection list
Name                         URI                                                         Identity                                                       Default     ReadWrite
podman-machine-default       ssh://core@127.0.0.1:53098/run/user/501/podman/podman.sock  /Users/edchong/.local/share/containers/podman/machine/machine  true        true
podman-machine-default-root  ssh://root@127.0.0.1:53098/run/podman/podman.sock           /Users/edchong/.local/share/containers/podman/machine/machine  false       true
```
 
 @webbnh @dbutenhof I was thinking we can add a podman remote test and execute `podman system connection list`, if there is a `podman-machine-default` use that. 
If there isn't one, skip the test. The user will have to set it up a `podman-machine-default` connection if they want to execute the test.

Comment (dbutenhof) [reply internal/cliwrapper/cliwrapper_test.go:55]: Webb pointed out that `podman` can create its own listener; so if the test environment can't run `systemctl --user`, 

```bash
podman system service --time=0 unix:///var/tmp/mypodman.sock &
podman system connection add test unix:///var/tmp/mypodman.sock
podman --connection=test ps
kill %1
```

It doesn't seem like there's a clean way to shut down the "system service" listener; just killing the backgrounded job seems to work OK.

File: internal/cliwrapper/cliwrapper_test.go [hunk 9]
```diff
@@ -48,6 +49,54 @@ func TestPodman_ImageExists(t *testing.T) {
...
+func TestPodman_Remote_ImageExists(t *testing.T) {
+	var tmpPodmanSocketCmd *exec.Cmd
+
+	// Check if there is an existing connection of `podman-machine-default` since this is included when installing
+	// podman desktop for macOS.
+	connectionName := "podman-machine-default"
+	chkDefaultConnectionCmd := exec.Command(tests.GetPodmanPath(), "--connection", connectionName, "system", "info") //nolint:gosec
+	if err := chkDefaultConnectionCmd.Run(); err == nil {
+		Podman_ImageExists(t, connectionName)
+	} else if runtime.GOOS == "linux" {
```
Comment (webbnh) [inline internal/cliwrapper/cliwrapper_test.go:65]: Apparently the linter thinks that the `else` clause is too complex for the embedded `err` construct.  Try changing this to
```go
	err := chkDefaultConnectionCmd.Run()
	if err == nil {
		Podman_ImageExists(t, connectionName)
	} else if runtime.GOOS == "linux" {
```

File: internal/cliwrapper/cliwrapper_test.go [hunk 10]
```diff
@@ -48,6 +49,54 @@ func TestPodman_ImageExists(t *testing.T) {
...
+	Podman_ImageExists(t, "")
+}
+
+func TestPodman_Remote_ImageExists(t *testing.T) {
+	var tmpPodmanSocketCmd *exec.Cmd
+
+	// Check if there is an existing connection of `podman-machine-default` since this is included when installing
+	// podman desktop for macOS.
+	connectionName := "podman-machine-default"
+	chkDefaultConnectionCmd := exec.Command(tests.GetPodmanPath(), "--connection", connectionName, "system", "info") //nolint:gosec
```
Comment (webbnh) [inline internal/cliwrapper/cliwrapper_test.go:62]: The linter would like to see you add `//nolint:gosec` to the _other_ `exec.Command()` invocations, as well.  I would appreciate it if you would add an additional comment, like:
```go"
	chkDefaultConnectionCmd := exec.Command(tests.GetPodmanPath(), "--connection", connectionName, "system", "info") //nolint:gosec  // Command line is trusted
```

File: internal/cliwrapper/cliwrapper_test.go [hunk 11]
```diff
@@ -48,6 +49,54 @@ func TestPodman_ImageExists(t *testing.T) {
...
+		addConnectionCmd := exec.Command(tests.GetPodmanPath(), "system", "connection", "add", connectionName, podmanSocketPath)
+		if err := addConnectionCmd.Run(); err != nil {
+			t.Fatalf("Failed to add connection: " + connectionName)
+		}
+
+		// Run test
+		Podman_ImageExists(t, connectionName)
+
+		// Clean up
+		if err := tmpPodmanSocketCmd.Process.Kill(); err != nil {
```
Comment (webbnh) [inline internal/cliwrapper/cliwrapper_test.go:86]: You should use a `defer()` function (or [a "cleanup" function](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addCleanup)) to ensure that the service process is cleaned up.

I'll toss you a PR.

**Review by webbnh — CHANGES_REQUESTED** [review]
Ed, this looks great!  Thanks again for doing this.

However, I think the test cleanup should be made more rigorous (I sent you a PR (https://github.com/ewchong/arcaflow-engine-deployer-podman/pull/2) for your branch), there are some lint complaints, and I have another suggestion (along with a few nits).

File: internal/cliwrapper/cliwrapper_test.go [hunk 12]
```diff
@@ -48,11 +49,59 @@ func TestPodman_ImageExists(t *testing.T) {
...
+		// The podman-machine-default doesn't exist then for Linux, create a temporary socket
+
+		// Setup
+		connectionName = "arcaflow-engine-deployer-podman-test"
+		podmanSocketPath := "unix:///var/tmp/" + connectionName + ".sock"
+
+		tmpPodmanSocketCmd = exec.Command(tests.GetPodmanPath(), "system", "service", "--time=0", podmanSocketPath)
+		if err := tmpPodmanSocketCmd.Start(); err != nil {
+			t.Fatalf("Failed to create temporary podman socket")
+		}
```
[the remark this answers is not in the document]
Comment (webbnh) [reply internal/cliwrapper/cliwrapper_test.go:75]: Nope, wrong, nevermind.  (I still have my Python hat on...sorry!)

File: internal/cliwrapper/cliwrapper_test.go (hunk 9 above)
Comment (dbutenhof) [reply internal/cliwrapper/cliwrapper_test.go:65]: You're simplifying the `if`, not the `else`. If the linter knows what it's talking about, the problem is the whole big Linux block below. You could instead stick with the `if` and simply return after doing the `Podman_ImageExists` then start a new `if runtime.GOOS == "linux"` block. Would that make it happier?

But, is it worth using the default Mac connection vs just always creating the temp service regardless of platform? (I think that "emergency cleanup" before the next run likely isn't an issue as the background service isn't being daemonized and ought to exit when the unit test process exits... but it wouldn't hurt to think about what happens if the socket or the connection definition already exists: for example, what if a system ends up running multiple copies at the same time? One solution would be to randomize the socket and connection names, for example by embedding a pid or a random UUID.)

File: internal/cliwrapper/cliwrapper_test.go [hunk 13]
```diff
@@ -48,11 +49,59 @@ func TestPodman_ImageExists(t *testing.T) {
...
+	if err := chkDefaultConnectionCmd.Run(); err == nil {
+		Podman_ImageExists(t, connectionName)
+	} else if runtime.GOOS == "linux" {
+		// The podman-machine-default doesn't exist then for Linux, create a temporary socket
+
+		// Setup
+		connectionName = "arcaflow-engine-deployer-podman-test"
+		podmanSocketPath := "unix:///var/tmp/" + connectionName + ".sock"
+
+		tmpPodmanSocketCmd = exec.Command(tests.GetPodmanPath(), "system", "service", "--time=0", podmanSocketPath)
```
Comment (dbutenhof) [inline internal/cliwrapper/cliwrapper_test.go:72]: Should we annotate with `//#nosec G204` to disable this annoying `gosec` message?

Comment (webbnh) [reply internal/cliwrapper/cliwrapper_test.go:72]: Yup, or with `//nolint:gosec`, but I like your more specific one, better (but existing practice seems to be the bigger hammer).

File: internal/cliwrapper/cliwrapper_test.go (hunk 11 above)
Comment (webbnh) [reply internal/cliwrapper/cliwrapper_test.go:86]: See https://github.com/ewchong/arcaflow-engine-deployer-podman/pull/2.

File: internal/cliwrapper/cliwrapper_test.go (hunk 9 above)
Comment (webbnh) [reply internal/cliwrapper/cliwrapper_test.go:65]: > Would that make it happier?

It's unhappy about the gross number of indentation levels (and, how it counts levels is slightly arcane).  In [my PR to Ed's branch](https://github.com/ewchong/arcaflow-engine-deployer-podman/pull/2), I ended up pulling most of the `else` body out into a helper function (because we needed to drop at least _two_ levels of indentation...).

> is it worth using the default Mac connection vs just always creating the temp service regardless of platform?

I think the challenge is figuring out how to create the temporary service on a platform other than `"linux"`.  Yeah, we can presumably do it on `"darwin"` _if_ Podman Desktop is installed, but it's much more straightforward to look for the "known"/"default" one and use it if it's there, and otherwise restrict us to where we know how to operate.  That, at least, lets things work in the CI.

Comment (webbnh) [reply internal/cliwrapper/cliwrapper_test.go:65]: > what happens if the socket or the connection definition already exists

Yeah, I didn't really like having the socket created at the top level of `/var/tmp`...but I eventually punted on that.  Having it created in the current working directory might be good.  Or, putting it under `/run/user/${UID}` (which is where Podman and Docker seem to put theirs) might be nice, but that's very linux-y.  But, at the least, giving it a random filename would be a very good idea.

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
👍   Thanks, again, Ed.

**Review by jaredoconnell — APPROVED** [review]
I find the specifier ConnectionName a little confusing, as it sounds more like an address to me, but other than that, looks good.

**Comment by ewchong** [comment]
> I find the specifier ConnectionName a little confusing, as it sounds more like an address to me, but other than that, looks good.

What you are referring to in Podman is called `URI`. In this case the `ConnectionName` represents the `name` column of the following table:
```
podman system connection list
Name                         URI                                                         Identity                                     Default     ReadWrite
podman-machine-default       ssh://core@localhost:55635/run/user/501/podman/podman.sock  /Users/cloud171/.ssh/podman-machine-default  true        false
podman-machine-default-root  ssh://root@localhost:55635/run/podman/podman.sock           /Users/cloud171/.ssh/podman-machine-default  false       false
```

The plugin itself doesn't do anything with the `URI` and has to be setup by the user beforehand.

**Comment by jaredoconnell** [comment]
> > I find the specifier ConnectionName a little confusing, as it sounds more like an address to me, but other than that, looks good.
> 
> What you are referring to in Podman is called `URI`. In this case the `ConnectionName` represents the `name` column of the following table:
> 
> ```
> podman system connection list
> Name                         URI                                                         Identity                                     Default     ReadWrite
> podman-machine-default       ssh://core@localhost:55635/run/user/501/podman/podman.sock  /Users/cloud171/.ssh/podman-machine-default  true        false
> podman-machine-default-root  ssh://root@localhost:55635/run/podman/podman.sock           /Users/cloud171/.ssh/podman-machine-default  false       false
> ```
> 
> The plugin itself doesn't do anything with the `URI` and has to be setup by the user beforehand.

Thanks for clarifying. Unfortunately that requires pre-setup before the workflow is run, but that's okay.

**Comment by webbnh** [comment]
> Unfortunately that requires pre-setup before the workflow is run, but that's okay.

We should be able to build on this functionality to add the ability for the workflow to set the URI.

It is pretty much always going to be the case that the Podman _listener_ has to be setup independent of the workflow, so some pre-setup will generally be required in order to use this feature, and I think that's totally reasonable.  However, I can see how it might be awkward to configure the Podman _connection_ outside of the workflow (depending on where and how the workflow is being run), in which case it would be handy to be able to provide it via the workflow.  (And, I don't think it would be hard.)
